### DISCUSSION TITLE
On exit Kitty kills? process(es) that had been nohup'ed and disown'ed

### DISCUSSION THREAD

**Discussion by ZuBB** [kovidgoyal/kitty#6544] [Q&A]
Env
* latest macOS
* latest kitty
* latest tmux
* latest fish shell

I have the next script that is being executed when I press Ctrl+D in the shell

```sh
function detach_logout_tasks
    for repo in (vcsh list)
        vcsh "$repo" commit -qam "autoupdate of configs in '$repo'" &>/dev/null
    end

    nohup vcsh push -q </dev/null &>/dev/null & disown
end

function on_exit3 --on-event fish_exit
    detach_logout_tasks
end
```

In short, it commits and pushes changes within dotfiles. This is supposed to happen in "background", despite the parent process being "gone".

when there are changes in dotfiles, this script works OK for the next mix of software
* Kitty + Tmux + Fish
* Terminal (default app) + Fish

But it does not work for Kitty + Fish case. Only commit phase succeeds and push results as undone

I played a bit with `close_on_child_death` option, but it did not help. 

Any ideas about what needs to be tweaked to have 3rd case also work?

**Comment by kovidgoyal** [comment]
This has nothing to do with kitty. Your attempt at putting a process
into the background is flawed in *various* ways. I suggest you google
how to daemonize processes on Unix. To truly decouple a process from its
environment you have to do a bunch of things like change the cwd to /,
replace the stdio file descriptors, call setsid() etc.

And that assumes that your process only talks to the tty via the stdio
file descriptors.

**Reply by ZuBB** [reply]
So cases \#1 and \#2 in a super strict and correct env also would not work. Am I right? and now they work by some coincidence?

**Comment by kovidgoyal** [comment]
If you don't daemonize properly whether a process survives its
controlling terminal being shutdown is undefined. 

For example, if the process is a python process, and you dont redirect
stdout/stderr to /dev/null a simple call to print() in the program will
raise an OSError and usually abort the program.

Similarly if you dont change the cwd to / various current dir related
operations can fail if the current dir changes it permissions or is
deleted.

If you dont call setsid() the process will receive signals sent to its
process group causing all kinds of problems.

On Sat, Aug 12, 2023 at 12:00:04AM -0700, Vasyl Zuziak wrote:
> So cases \#1 and \#2 in a super strict and correct env also would not work. Am I right? and now they work by some coincidence?
> 
> -- 
> Reply to this email directly or view it on GitHub:
> https://github.com/kovidgoyal/kitty/discussions/6544#discussioncomment-6707453
> You are receiving this because you commented.
> 
> Message ID: ***@***.***>

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Reply by ZuBB** [reply]
I admit that you are 100% right about what needs to be done for processes to be left in the background and run properly. But I still can't wrap my head around the fact that my old setup survived for smth like 7 years through Linux & mac boxes, different terminals, and shells, and all was OK, but suddenly it refused to work with one particular case with kitty

Nevertheless search bring me to the next links
* https://unix.stackexchange.com/questions/266565/daemonize-a-process-in-shell
* https://software.clapper.org/daemonize/

And I was able to solve my task with `daemonize` CLI app

**Comment by Ihirasd** [comment]
oh its true
